docs/core/concepts/results.md

```diff
@@ -28,44 +20,116 @@ An instantiated Result object has the following attributes:
+```
+
+In Core, the return value is associated in-memory with the `Result` object attached to flow `State`s. For Cloud customers, the return value of a result handler's `write` method is also persisted in Prefect Cloud's database.
 
 ::: warning Handle your data carefully
-When running on Prefect Cloud, the output of a Result Handler's `write` method is what is stored in the Cloud database.
+Using Result Handlers means that data will be persisted beyond the Prefect's Python process to a storage location, so it is worth taking some extra time to consider what data is safe to persist where.
+
+When running on Prefect Cloud, the return value of a Result Handler's `write` method is what is stored in the Cloud database. Though you may serialize your data during your result handler's execution, *only* `return` data from Tasks that you feel is safe to persist in Prefect Cloud.
 :::
 
 ### How to specify a `ResultHandler`
 
-There is a hierarchy to determining what `ResultHandler` to use for a given piece of data:
+`ResultHandler`'s `read` and `write` methods will be called by Prefect only if you enable checkpointing. Once this configuration is set, there is a hierarchy to determining what `ResultHandler` to use for a given piece of data:
 
-1. First, users can set a global default in their Prefect user config; if you never mention or think about Result Handlers again, this is the handler that will always be used.
-1. Next, you can specify a Flow-level result handler at Flow-initialization using the `result_handler` keyword argument. Once again, if you never specify another result handler, this is the one that will be used for all your tasks in this particular Flow.
+1. You can specify a Flow-level result handler at Flow-initialization using the `result_handler` keyword argument.If you never specify another result handler, this is the one that will be used for all your tasks in this particular Flow.
 1. Lastly, you can set a Task-level result handler. This is achieved using the `result_handler` keyword argument at Task initialization (or in the `@task` decorator). If you provide a result handler here, it will _always_ be used if the _output_ of this Task needs to be cached for any reason whatsoever.
 
 ::: tip The Hierarchy
-Task-level result handlers will _always_ be used over Flow-level handlers; Flow-level result handlers will _always_ be used over the handler set in your config.
+Task-level result handlers will _always_ be used over Flow-level handlers. Neither will be used if a Task's `checkpointing` kwarg is set to `False`, or the global `prefect.config.tasks.checkpointing` value is set to `False`.
+:::
+
+::: warning Result Handlers are always attached to Task outputs
+For example, suppose Task A has Result Handler A, and Task B has Result Handler B, and that A passes data downstream to B. If B fails and requests a retry, it needs to cache its inputs, one of which came from A. When it stores this data, it will use Result Handler A.
```

Comment:
```suggestion
For example, suppose task A has result handler A, and task B has result handler B, and that A passes data downstream to B. If B fails and requests a retry, it needs to cache its inputs, one of which came from A. When it stores this data, it will use result handler A.
```